diff --git a/client/listener.go b/client/listener.go
index 04f9e4c..e061939 100644
--- a/client/listener.go
+++ b/client/listener.go
@@ -645,6 +645,9 @@ func NewListener(client *Client, rates *Rates, egressCh chan *Request, logBacken
 	}
 
 	l.decoySender = newSender(l.PickNextRequest, egressCh, client.cfg.Debug.DisableDecoyTraffic, logBackend)
+	if rates != nil {
+		l.decoySender.UpdateRates(rates)
+	}
 
 	l.listener, err = client.cfg.Listen.Listen()
 	if err != nil {
diff --git a/client/sender.go b/client/sender.go
index e66cd53..60aba42 100644
--- a/client/sender.go
+++ b/client/sender.go
@@ -135,7 +135,7 @@ func (s *sender) UpdateConnectionStatus(isConnected bool) {
 }
 
 func (s *sender) UpdateRates(rates *Rates) {
-	if rates.messageOrLoop <= 0 {
+	if rates == nil || rates.messageOrLoop <= 0 {
 		s.log.Warning("Invalid messageOrLoop rate, ignoring")
 		return
 	}
